### PR TITLE
bpo-42115: Refactor the opcode cache into its own file

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue42115

The advantage of doing this is that the code in `ceval.c` is much easier to manage now.

### CODE REVIEW

File: Python/opcode_cache.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    {
+        return 0;
+    }
+
+    _PyOpcache_LoadGlobal *lg = &co_opcache->u.lg;
+
+    if (co_opcache->optimized > 0 &&
+        lg->globals_ver == ((PyDictObject *)f->f_globals)->ma_version_tag &&
+        lg->builtins_ver == ((PyDictObject *)f->f_builtins)->ma_version_tag) {
+
```
Comment (1st1) [inline Python/opcode_cache.c]:
```suggestion
        lg->builtins_ver == ((PyDictObject *)f->f_builtins)->ma_version_tag
    ) {

```

File: Python/ceval.c [hunk 2]
```diff
@@ -2657,68 +2494,22 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
...
 
-                    co_opcache->optimized = 1;
-                    lg->globals_ver =
-                        ((PyDictObject *)f->f_globals)->ma_version_tag;
-                    lg->builtins_ver =
-                        ((PyDictObject *)f->f_builtins)->ma_version_tag;
-                    lg->ptr = v; /* borrowed */
+            OPCACHE_CHECK();
+            if (_PyOpcodeCache_LoadGlobal(co, &co_opcache, f, name, &v) != 0) {
+                if (!_PyErr_OCCURRED()) {
```
Comment (1st1) [inline Python/ceval.c]: I don't like this check here. I'd communicate it via the return value better.

File: Python/opcode_cache.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    return _PyOpcodeCache_ReloadLoadGlobal(co, co_opcache_ptr, f, name, result);
+}
+
+
+
+static int _PyOpcodeCache_InitializeLoadAttr(PyCodeObject *co, _PyOpcache **co_opcache_ptr,
+                                             PyTypeObject* type, PyObject* name,
+                                             PyObject* owner, PyObject** result) {
+
```
Comment (1st1) [inline Python/opcode_cache.c:134]:
```suggestion
                                             PyObject* owner, PyObject** result
) {

```

File: Python/opcode_cache.h [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+extern size_t opcache_attr_opts;
+extern size_t opcache_attr_hits;
+extern size_t opcache_attr_misses;
+extern size_t opcache_attr_deopts;
+extern size_t opcache_attr_total;
+#endif
+
+/* macros for opcode cache */
+#define OPCACHE_CHECK() \
```
Comment (1st1) [inline Python/opcode_cache.h]: These no longer have to be a macro. They can be inlinable static funcs.

**Comment by 1st1** [comment]
Overall +1. @pablogsal please copy me in opcode cache PRs ;)

File: Python/opcode_cache.h (hunk 4 above)
Comment (pablogsal) [reply Python/opcode_cache.h]: :+1: Good idea, will convert them!

**Comment by pablogsal** [comment]
I checked and we lose quite a bit of performance (3-4%) due to the extra function calls. I think I am going to close this for now :(

**Pull request closed without merging**